### DISCUSSION TITLE
LDAP Example

### DISCUSSION THREAD

**Discussion by joaobrazuna** [Leantime/leantime#694] [General]
Is there any tutorial (with examples) to how to setup LDAP on leantime?

In the  file /config/configuration.php there is the following reference:
  /*ldap default settings (can be changed in company settings) */
How can this setting be activated/showed?

Many Thanks.

**Comment by Treuman** [comment]
The question is relevant. You may need documentation for an update

**Comment by marcelfolaron** [comment]
Hi All,

I just added a quick overview in https://docs.leantime.io/#/installation/configuration?id=ldap-configuration-beta
Please let me know if that helps or if you have any other questions.

**Reply by joaobrazuna** [reply]
Many thanks for the quick overview, but I still need help.

I'm trying to connect to a Windows Server AD from a Oracle Linux, and I'm using the following json for $ldapKeys:
 	public $ldapKeys = '{ 
        "username":"sAMAccountName",
        "groups":"memberof",
        "email":"mail",
        "firstname":"givenname",
        "lastname":"sn"
        }';

The error reported is:
Warning: ldap_search(): Search: Operations error in /var/www/html/leantime/src/domain/ldap/services/class.ldap.php on line 192
Unable to search LDAP server

I can browse all the users on the correct OU when I run a LDAP Browser on the server.

-Which format should I use for $bindUser?
"cn=leantimeuser,DC=example,DC=com" <- is this correct? 
or just " $bindUser = "leantimeuser"

-In the file /config/configuration.php there is the following reference:
/*ldap default settings (can be changed in company settings) */
How can this setting be activated/showed?

-Where can I indicate the LDAP filter in order to only retrieve the users of a certain security group?

note: I tried the same settings I use on other opensource (https://gitea.io/) to connect to the same Windows Server AD, with no luck. 

Many thanks in advance for your help.

**Comment by marcelfolaron** [comment]
Oh, I also just noticed that this error comes up in the import (which is
currently not fully implemented).
Can you login with your ldap config?

**Reply by joaobrazuna** [reply]
I can not.
I tried to login directly on the login page, but reports always that the user does not exists or password is wrong.

**Comment by Treuman** [comment]
Hi @marcelfolaron 
I was unable to connect to LDAP AD, Apache reports a 408 error (timeout)
Port 389 forwarded
Used AD LDAP (microsoft windows)
Created groups in AD: "leantime-administrators" and "leantime-users" - added users inside
Mapped keys leantime:LDAP
What other solutions will be available?
This is my configuration.php:

```
/*ldap default settings (can be changed in company settings */
    /*ldap default settings*/
    public $useLdap = true;
    public $ldapHost = "domain.my"; //FQDN
    public $ldapPort = 389; //Default Port
    public $baseDn = "DC=domain,DC=my"; //Base DN, example: DC=example,DC=com
    public $ldapDn = "CN=Users,OU=Organization,DC=domain,DC=my"; //Location of users, example: CN=users,DC=example,DC=com
    public $ldapUserDomain = "@domain.my"; //Domain after ldap, example @example.com
    public $bindUser = "userAuth@domain.my"; //ldap user that can search directory. (Should be read only)
    public $bindPassword = "password";

    //Default ldap keys in your directory. 
    public $ldapKeys = '{
        "username":"sAMAccountName",
        "groups":"memberof",
        "email":"mail",
        "firstname":"giveName",
        "lastname":"sn"
        }';

    //Default role assignments upon first login. (Optional) Can be updated in user settings for each user
    public $ldapLtGroupAssignments = '{
          "10": {
            "ltRole":"client",
            "ldapRole":"leantime-users" //Group from AD (LDAP)
          },
          "20": {
            "ltRole":"developer",
            "ldapRole":""
          },
          "30": {
            "ltRole":"clientManager",
            "ldapRole":""
          },
          "40": {
            "ltRole":"manager",
            "ldapRole":""
          },
          "50": {
            "ltRole":"admin",
            "ldapRole":"leantime-administrators" //Group from AD (LDAP)
          }
        }';
```

thx

**Comment by marcelfolaron** [comment]
Let me do some additional testing on Windows AD. The config looks correct.
The timeout would indicate that something is not correct with the baseDN
(or firewall rules)

**Comment by marcelfolaron** [comment]
Could you all try something for me before I set up a new ldap server. Change this line

https://github.com/Leantime/leantime/blob/e8376104b1f44a22e2a7ffa8717302dd80ace086/src/domain/ldap/services/class.ldap.php#L126

to 
`return @ldap_bind($this->ldapConnection, $username."@".$this->host, $passwordBind);`

**Reply by Treuman** [reply]
> Could you all try something for me before I set up a new ldap server. Change this line
> 
> https://github.com/Leantime/leantime/blob/e8376104b1f44a22e2a7ffa8717302dd80ace086/src/domain/ldap/services/class.ldap.php#L126
> 
> to `return @ldap_bind($this->ldapConnection, $username."@".$this->host, $passwordBind);`

Thank you, when the line is changed, the authorization error is also

**Reply by joaobrazuna** [reply]
Hi,
The changes you told, do not solve the problem, but if I change the same line to 
126           **return @ldap_bind($this->ldapConnection, $bindUser , $bindPassword );**

it works.
After this change, all AD users can now authenticate with AD password.


But, now the option  "User management" does not work and always reports the error:
Warning: ldap_search(): Search: Operations error in /var/www/html/leantime/src/domain/ldap/services/class.ldap.php on line 194
Unable to search LDAP server

**Comment by marcelfolaron** [comment]
Alright, I didn't realize it at first but Active Directory is using a different DN to authenticate a user. 
The code was written to support OpenLdap. I will add that to the docs. I'll work on AD support, should be pretty quick but it will have to be passed in as a config variable

**Reply by Treuman** [reply]
Hey @marcelfolaron .
How is the project LDAP Active Directory progressing? Waiting for a quick decision or instructions on how to proceed.
thx

**Comment by marcelfolaron** [comment]
Oh interesting. Where are those variables coming from?
Is it $this->bindUser ?

**Comment by marcelfolaron** [comment]
I'll check if that works for openldap as well. If not we might just add a
variable for ldapType

**Reply by joaobrazuna** [reply]
My mistake
it only works  when the line 126 is:

return @ldap_bind($this->ldapConnection, "userAuth@domain.my", "password" );

**Comment by joaobrazuna** [comment]
Hi there,
Back to LDAP-AD integration.
Tried it again, but cannot put it to work again.

Is there any log where to check what is going wrong?
Many Thanks

**Comment by marcelfolaron** [comment]
I made a few updates in the last version to add more logging. Please try
that and look for any issues in resources/logs/error.log

**Comment by joaobrazuna** [comment]
Nope,
In that log file, only the following 1 line:

[02-Aug-2022 14:48:39 Europe/Lisbon] PHP Deprecated:  Constant FILTER_SANITIZE_STRING is deprecated in /var/www/html/leantime/src/domain/api/controllers/class.sessions.php on line 74

**Reply by marcelfolaron** [reply]
Couple things then:
1. Make sure the config var `$useLdap` is set to `true`.
2. Please make sure you have the php ext-ldap extension installed.

**Comment by joaobrazuna** [comment]
ext-ldap was missing.
Now it "almost" works.
It cannot retrieve the firstname from the AD (check #1 below)
![image](https://user-images.githubusercontent.com/44297535/182401313-054cc938-fe0c-4136-b32b-ba89c2e487ba.png)

The $ldapKeys is correct
![image](https://user-images.githubusercontent.com/44297535/182399831-242bfde4-d071-4ec8-a5b0-9161041e75ad.png)



by the way,,, another messages missing translations (check #2)

**Comment by marcelfolaron** [comment]
So for some odd reason I noticed that it works when the ldap key is
actually all lowercase "givenname"
Had another person reach out with the same issue and that fixed it. Not
sure why that is happening yet.

**Comment by marcelfolaron** [comment]
Hi,

Thanks for the feedback. Bear with me as we figure that out.
$bindUser should only be the username so "leantimeuser" would be correct.

I tested it with the synology ldap server.

The setting that determines which folder to look for would be in $ldapDn:

"cn=Users,dc=192,dc=168,dc=1,dc=2";

So `cn` should be set to the folder of your users.

Do you see any other errors?

**Comment by marcelfolaron** [comment]
Got it. Yes, I believe this only works with Active Directory and not with
OpenLdap.
Thank you for your support. I'll add those changes.

**Comment by joaobrazuna** [comment]
Hi 
V.2.2.10 is working 100% now with Active directory

1# Adicional Feature Request
Like other platforms do (eg. gitea), it would be nice if a AD user could login only with AD username.
Example: 
Besides the user doing the login like this:->  name.surname@domain.local
Also be possible to login like this:-> name.surname
i.e. without having to include the domain.
Meaning, that Leantime with AD active would first query the AD, and only if the query is negative, would query the internal database.

2# Adicional Feature Request
Possibility to put "$ldapDefaultRoleKey = 0", and if no "ldapRole" is met, the new login would be refused.
The propose is: on an AD with lots of users, only the ones belonging to at least one security group would be accepted to login, and not follow the $ldapDefaultRoleKey value, and thus not allowing for a new user to be created automatically.
True that now it would be created with no access to anything, but I think this way would prevent "curious" users to try to login.

Many thanks

**Comment by Treuman** [comment]
### working version to 2.2.10
**First name and last name displayed**

/*ldap default settings (can be changed in company settings) */
    public $useLdap = true; 
    public $ldapType = 'AD';
    public $ldapHost = 'mydomain.ru';
    public $ldapPort = 389; 
    public $baseDn = 'DC=mydomain,DC=ru';
    public $ldapDn = 'OU=Users,OU=Organization,DC=mydomain,DC=ru'
    public $ldapUserDomain = '@mydomain.ru';
    public $bindUser = 'login_user_from_ad';
    public $bindPassword = 'password_user_from_ad';
    public $ldapKeys = '{
        "username":"samaccountname",
        "groups":"memberof",
        "email":"mail",
        "firstname":"givenname",
        "lastname":"sn",
        "phonenumber":"telephonenumber"   //not displayed
        }';

**Comment by lockdlock** [comment]
In version 2.3.1 beta, working AD login
I would expect two
1.be able to log in with my username without a domain.
2.Ability to customise priorities.

**Comment by Mrs-Feathers** [comment]
i also suggest the ability to login without an @domain. all users of mine are LDAP users period, so its redundant and because corporate has a different network, the domain is different from our emails so we have to login with emailusername@verystrangedomain.com and its awkward to have to remind everyone "its different. we dont know why its different, it just is" but otherwise its an amazing and useful platform for tracking team projects

**Reply by TimexPeachtree** [reply]
Yes this is a very good option, required especially in organizations where not all employees have corporate email address.
